src/components/JustNotSorry.js

```diff
@@ -18,74 +19,66 @@ const WATCH_FOR_NEW_NODES = {
   childList: true,
 };
 
-class JustNotSorry extends Component {
-  constructor(props) {
-    super(props);
-    this.state = {
-      parentNode: {},
-      warnings: [],
-    };
+const JustNotSorry = ({ onEvents }) => {
+  const [state, setState] = useState({ warnings: [], parentNode: {} });
 
-    this.resetState = this.resetState.bind(this);
-    this.handleSearch = this.handleSearch.bind(this);
-    this.updateWarnings = this.updateWarnings.bind(this);
-    this.applyEventListeners = this.applyEventListeners.bind(this);
+  const resetState = () => {
+    setState({ warnings: [], parentNode: {} });
+  };
 
-    this.documentObserver = new MutationObserver(
-      forEachUniqueContentEditable(this.applyEventListeners)
-    );
-    this.documentObserver.observe(document.body, WATCH_FOR_NEW_NODES);
-  }
-
-  resetState() {
-    this.setState({ parentNode: {}, warnings: [] });
-  }
-
-  updateWarnings(email, patterns) {
+  const updateWarnings = (email, patterns) => {
     if (!email || !email.offsetParent) {
-      this.resetState();
-      return;
+      return resetState();
     }
+
     const newWarnings =
       email.childNodes.length > 0
         ? Array.from(email.childNodes)
-            .filter((node) => node.textContent !== '')
-            .flatMap((text) => findRanges(text, patterns))
+            .filter((n) => n.textContent !== '')
+            .flatMap((n) => findRanges(n, patterns))
         : findRanges(email, patterns);
 
-    this.setState(({ parentNode }) =>
-      parentNode.id !== email.offsetParent.id
-        ? { parentNode: email.offsetParent, warnings: newWarnings }
-        : { parentNode, warnings: newWarnings }
-    );
-  }
```

Comment:
Using the [updater function form](https://beta.reactjs.org/reference/react/useState#updating-state-based-on-the-previous-state) of `setState` was a performance fix for long emails with a lot of warnings.  The updater function form can be slow, but the version that takes an object seems to be even slower.  I'm not entirely sure why.

To reproduce, try pasting the following paragraph into an email 10 times:

> I'm sorry about being so sorry.  I'm just not sorry about being sorry.  Where is it?  Does that make sense? Oh, I'm just sorry.  So much sorry.  Is that okay? Do you mind? Forgive me for trying to be yoda.  I'm no expert, but I'm sorry about being sorry.  I'm actually just sorry, very, very sorry.  I kind of want to know, does this make sense? I feel that I believe honestly that this might be a stupid question.

and then click in and out of the compose window to trigger the warnings to re-render.

Reply:
I dug into things more and found the root cause of the slowness in the new code.  It's due to how the MutationObserver is being created:

```js
  const documentObserver = new MutationObserver(
    forEachUniqueContentEditable(applyEventListeners)
  );
  documentObserver.observe(document.body, WATCH_FOR_NEW_NODES);
```

Each time the component re-renders, a new MutationObserver is created and told to observe the document body.  The more stuff that is typed into the compose window (particularly line breaks), the more times `applyEventListeners` is being run and the more duplicate events are being triggered.

Inspired by [this article](https://blog.logrocket.com/guide-to-custom-react-hooks-with-mutationobserver/), I tried wrapping the MutationObserver logic in useEffect and storing the observer in the state and that fixed the problem:
```js
  const [observer, setObserver] = useState(null);
  // ...
  // function definitions here
  // ...
  useEffect(() => {
    if (observer) return;

    const callback = forEachUniqueContentEditable(applyEventListeners);
    const obs = new MutationObserver(callback);
    obs.observe(document.body, WATCH_FOR_NEW_NODES);
    setObserver(obs);
    return () => {
      if (observer) {
        console.log("disconnecting observer")
        observer.disconnect();
      }
    };
  }, [forEachUniqueContentEditable, applyEventListeners, setObserver, observer]);
```

Reply:
Thank you for looking into this and catching it! I went on vacation and opened this pr a little prematurely- Maybe in the future a test can be added that validates the performance is not affected by a change.

Reply:
Good idea.  That would be a useful test to have.  I wonder if we could do it with jsdom or if we'd need to tie into chrome?

Reply:
I manually tested the new code with two scenarios:

1. Keep adding paragraphs of text without any warning phases
2. Keep adding paragraphs of text containing multiple warning phrases

After each addition, I checked for responsiveness by trying to move the cursor to a different location and by clicking the subject field and back into the email again (to trigger the focus event).

In Scenario 1, the performance stayed consistently good no matter how many paragraphs I added.

In Scenario 2, the performance got steadily worse the more paragraphs I added.  This is consistent with the behavior of the existing v2.5 code.

Our automated test should cover both of these scenarios.  Ideally, the performance should be consistent no matter how many warning phrases appear.

Reply:
I created an issue to track this: https://github.com/defmethodinc/just-not-sorry/issues/179